src/cryptonote_core/tx_pool.cpp

```diff
@@ -411,17 +424,28 @@ namespace cryptonote
       MINFO("Pool weight after pruning is larger than limit: " << m_txpool_weight << "/" << bytes);
   }
   //---------------------------------------------------------------------------------
-  bool tx_memory_pool::insert_key_images(const transaction_prefix &tx, const crypto::hash &id, bool kept_by_block)
+  bool tx_memory_pool::insert_key_images(const transaction_prefix &tx, const crypto::hash &id, relay_method tx_relay)
   {
     for(const auto& in: tx.vin)
     {
       CHECKED_GET_SPECIFIC_VARIANT(in, const txin_to_key, txin, false);
       std::unordered_set<crypto::hash>& kei_image_set = m_spent_key_images[txin.k_image];
-      CHECK_AND_ASSERT_MES(kept_by_block || kei_image_set.size() == 0, false, "internal error: kept_by_block=" << kept_by_block
-                                          << ",  kei_image_set.size()=" << kei_image_set.size() << ENDL << "txin.k_image=" << txin.k_image << ENDL
-                                          << "tx_id=" << id );
-      auto ins_res = kei_image_set.insert(id);
-      CHECK_AND_ASSERT_MES(ins_res.second, false, "internal error: try to insert duplicate iterator in key_image set");
+
+      bool visible = true;
+      if (tx_relay != relay_method::block)
+      {
+        for (const crypto::hash& other_id : kei_image_set)
+          visible &= m_blockchain.txpool_tx_matches_category(other_id, relay_category::broadcasted);
+      }
+
+      CHECK_AND_ASSERT_MES(visible, false, "internal error: tx_relay=" << unsigned(tx_relay)
+                                           << ", kei_image_set.size()=" << kei_image_set.size() << ENDL << "txin.k_image=" << txin.k_image << ENDL
+                                           << "tx_id=" << id);
+
+      const bool added =
+        kei_image_set.insert(id).second ||
+        m_blockchain.txpool_tx_matches_category(id, relay_category::broadcasted);
+      CHECK_AND_ASSERT_MES(added, false, "internal error: try to insert duplicate iterator in key_image set");
```

Comment:
I will add a comment here and above. Yes, the intent is to accept a duplicate key-image if the tx originated at the node or was received via stem. Once its received by any other method, it should switch to standard mode. If the tx is received twice over stem, it will be switched to "fluff" elsewhere. This check should pass until its fluffed.

Reply:
I couldn't come up with a better name btw. "Broadcast" is intended to mean "send to many participants" opposed to "limited set or none".

Reply:
If the tx originated at the node, you should not receive a duplicate key image unless the user tries to double spend, in which case you likely want to error out. It's the right thing to do for stem though, so these two cases need separating. I think broadcast is a good name. If you're looking for another, maybe public or unrestricted.

Reply:
When writing the comments I realized that there were bugs in both portions. The boolean logic was inverted what it needed to be. The core tests I wrote didn't catch this because of a second bug in the mempool `add_tx` function. Yikes.

Please re-review this, and all changes in tx_pool.